File: pddapp/notifiers/webhook.py
"""Webhook notifier — posts a JSON payload to a configured URL.

The payload includes a ``content`` field so it works out-of-the-box with Discord
and Slack incoming webhooks, plus structured fields for custom consumers.
"""

from __future__ import annotations

import logging

import httpx

from .base import Alert, Notifier

logger = logging.getLogger("pddapp.alerts")


class WebhookNotifier(Notifier):
    name = "webhook"

    def __init__(self, url: str, *, timeout: float = 10.0) -> None:
        self.url = url
        self.timeout = timeout

    def send(self, alert: Alert) -> None:
        if not self.url:
            logger.warning("Webhook enabled but WEBHOOK_URL is empty; skipping.")
            return

        payload = {
            # Discord/Slack-friendly:
            "content": f"{alert.subject}\n{alert.url}".strip(),
            "text": alert.subject,
            # Structured fields for custom consumers:
            "alert": {
                "product_name": alert.product_name,
                "goods_id": alert.goods_id,
                "url": alert.url,
                "price": alert.price,
                "threshold_price": alert.threshold_price,
                "original_price": alert.original_price,
                "currency": alert.currency,
            },
        }
        resp = httpx.post(self.url, json=payload, timeout=self.timeout)
        resp.raise_for_status()

File: src/lib/notifiers/webhook.ts
// Webhook notifier — posts JSON to a configured URL. The `content` field makes
// it work out-of-the-box with Discord and Slack incoming webhooks; structured
// fields are included for custom consumers.

import { Alert, Notifier, alertSubject } from "./types";

export class WebhookNotifier implements Notifier {
  readonly name = "webhook";

  constructor(private readonly url: string) {}

  async send(alert: Alert): Promise<void> {
    if (!this.url) return;
    const payload = {
      content: `${alertSubject(alert)}\n${alert.url}`.trim(),
      text: alertSubject(alert),
      alert,
    };
    const res = await fetch(this.url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      throw new Error(`webhook responded ${res.status}`);
    }
  }
}
